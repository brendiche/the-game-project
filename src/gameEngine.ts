import { CharacterProperties, Target } from "./helper";

type CallbackGameEngine = () => void
export class Engine {
  private readonly runnigThread: CallbackGameEngine[];
  private character: CharacterProperties;
  private targets: Target[] = [];

  constructor(refreshTimer = 20){
    this.runnigThread = [];
    setInterval(() => {
      for (const callback of this.runnigThread) {
        callback();
      }
      if(this.character && this.targets.length){
        for (const target of this.targets) {
          if(this.character.position === target.position) alert('colision detected')
        }
      }
    }, refreshTimer)
  }

  public addGamingThread(callback: CallbackGameEngine): void {
    this.runnigThread.push(callback);
  }

  public setCharacter(character: CharacterProperties){
    this.character = character;
  }

  public addTarget(target: Target){
    console.log('[gameEngine][addTarget] this',this)
    this.targets.push(target);
  }
}