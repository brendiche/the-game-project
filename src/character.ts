import { addGamingThread } from './gameEngine';
import { getPosition, SideType } from './helper';
import { createItem } from './item';
import './Naruto.css'

const OFFSET_CHARACTER = {
  top: 12,
  left: 55,
}
const States = ['stand', 'run', 'jump', 'throw', 'down'] as const;
type StateType = typeof States[number];
let state: StateType = 'stand';
let side: SideType = 'right';

export const Character = (): HTMLElement => {
  const component = document.createElement('div');
  component.className = `naruto-${state}`;
  addGamingThread(() => test(component));
  addListeners(component);
  return component;
}

const test = (component: HTMLElement) => {
  setState(component, state);
  setSide(component, side);
}

const setState = (component: HTMLElement, state: StateType): void => {
  // console.log('[character][setState] class :', component.className);
  const classRegex = new RegExp(`naruto-(${States.join('|')})`);
  component.className = component.className.replace(classRegex, `naruto-${state}`);
}

const setSide = (component: HTMLElement, side: SideType): void => {
  if(side === 'left'){
    component.className= component.className.includes('left') ? component.className : `${component.className} left`;
  }else{
    component.className = component.className.replace('left','');
  }
}

const addListeners = (component: HTMLElement): void => {
  window.addEventListener('keydown' , (event) => {
    console.log('[character][addListeners] keydown: ', event.key);
     switch(event.key){
       case 'ArrowRight':
         state = 'run';
         side = 'right';
       break;
       case 'ArrowLeft':
          state = 'run';
          side = 'left';
        break;
       case 'ArrowDown':
          state = 'down';
        break;
       case 'ArrowUp':
         state = 'jump';
        setTimeout(() => {
          if(state === 'jump') {
            state = 'stand';
          }
        },701)
        break;
       case ' ':
          state = 'throw';
          createItem('kunai', {
            position: 'absolute',
            left: `${getPosition(component) + OFFSET_CHARACTER.left}px`,
            top: `${getPosition(component, 'top') + OFFSET_CHARACTER.top}px`,
          });
          setTimeout(() => {
            if(state === 'throw'){
              state = 'stand';
            }
          },501)
          break;
          
       case 's':
          state = 'throw';
          createItem('shuriken', {
            position: 'absolute',
            left: `${getPosition(component) + OFFSET_CHARACTER.left}px`,
            top: `${getPosition(component, 'top') + OFFSET_CHARACTER.top}px`,
          });
          setTimeout(() => {
            if(state === 'throw'){
              state = 'stand';
            }
          },501)
          break;
          
     }
  });
  window.addEventListener('keyup' , (event) => {
    console.log('[character][addListeners] keyup: ', event.key);
     switch(event.key){
        case 'ArrowRight':
        case 'ArrowLeft':
        case 'ArrowDown':
          state = 'stand';
        break;
     }
  });
}
