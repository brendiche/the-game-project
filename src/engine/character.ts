import { Engine } from './gameEngine';
import { CharacterConfig, SideType, StatesRPG, StatesRPGType, StateType } from './helper';

export class Character {
  private readonly htmlElement: HTMLElement;
  private readonly characterName: string;
  private readonly engine: Engine;
  private readonly config: CharacterConfig;
  private _state: StateType = 'stand';
  private _stateRPG: StatesRPGType = 'stand';
  private _side: SideType = 'right';

  constructor(config: CharacterConfig, engine: Engine, characterName: string){
    this.config = config;
    this.engine = engine;
    this.characterName = characterName;
    this.htmlElement = document.createElement('div');
    this.htmlElement.className = `${characterName}-${this._state}`
    this.engine.addGamingThread(() => this.engineCallback());
    this.addListeners(); 
  }
  get state(): StateType{
    return this._state;
  }
  get stateRPG(): StatesRPGType{
    return this._stateRPG;
  }
  get side(): SideType{
    return this._side;
  }
 
  get element(): HTMLElement{
    return this.htmlElement;
  }

  private setState(): void {
    const classRegex = new RegExp(`${this.characterName}-(${StatesRPG.join('|')})`);
    this.htmlElement.className = this.htmlElement.className.replace(classRegex, `${this.characterName}-${this._stateRPG}`);
  }
  

  private engineCallback(): void {
    this.setState();
  }

  private addListeners():void{
    window.addEventListener('keydown', (event) => {
      switch(event.key){
        case 'ArrowRight':
          this._stateRPG = 'right';
          break;
        case 'ArrowLeft':
          this._stateRPG = 'left';
          break;
        case 'ArrowDown':
          this._stateRPG = 'down';
          break;
        case 'ArrowUp':
          this._stateRPG = 'top';
          break;
      }
    });
    window.addEventListener('keyup' , (event) => {
      console.log('[character][addListeners] keyup: ', event.key);
       switch(event.key){
          case 'ArrowRight':
          case 'ArrowLeft':
          case 'ArrowDown':
          case 'ArrowUp':
            this._stateRPG = 'stand';
          break;
       }
    });
  }
}
