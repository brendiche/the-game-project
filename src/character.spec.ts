/**
 * @jest-environment jsdom
 */
import { Character } from "./character";
import { Engine } from "./gameEngine";
const dateNowStub = () => 1644182219303;
global.Date.now = dateNowStub;


const mockGamingThreadCallback = jest.fn();
const mockEngine: Partial<Engine> = {
  addGamingThread: mockGamingThreadCallback,
}
const character = new Character(mockEngine as Engine, 'test');
document.body.appendChild(character.element);

describe('charater class', () => {
  it('it should create a character well initialized', () => {
    expect(character.element).toMatchSnapshot();
    expect(character.properties).toMatchSnapshot();
    expect(mockGamingThreadCallback).toHaveBeenCalled();
  });
  it('should test the KeyboardEvent keydown', () => {
    const arrowDownEvent = new KeyboardEvent('keydown', {key: 'ArrowDown'});
    const arrowLeftEvent = new KeyboardEvent('keydown', {key: 'ArrowRight'});
    const arrowRightEvent = new KeyboardEvent('keydown', {key: 'ArrowLeft'});
    const arrowUpEvent = new KeyboardEvent('keydown', {key: 'ArrowUp'});
    const spaceEvent = new KeyboardEvent('keydown', {key: ' '});
    const sEvent = new KeyboardEvent('keydown', {key: 's'});
    window.dispatchEvent(arrowDownEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(arrowLeftEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(arrowRightEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(arrowUpEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(spaceEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(sEvent);
    expect(character.properties).toMatchSnapshot();
  }); 
  it('should test the KeyboardEvent keyup', () => {
    const arrowDownEvent = new KeyboardEvent('keyup', {key: 'ArrowDown'});
    const arrowLeftEvent = new KeyboardEvent('keyup', {key: 'ArrowRight'});
    const arrowRightEvent = new KeyboardEvent('keyup', {key: 'ArrowLeft'});
    window.dispatchEvent(arrowDownEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(arrowLeftEvent);
    expect(character.properties).toMatchSnapshot();
    window.dispatchEvent(arrowRightEvent);
    expect(character.properties).toMatchSnapshot();
  })
})