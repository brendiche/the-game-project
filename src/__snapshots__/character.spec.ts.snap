// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`charater class it should create a character well initialized 1`] = `
<div
  class="test-stand"
/>
`;

exports[`charater class it should create a character well initialized 2`] = `
Object {
  "items": Array [],
  "position": NaN,
  "side": "right",
  "state": "stand",
}
`;

exports[`charater class should test the KeyboardEvent keydown 1`] = `
Object {
  "items": Array [],
  "position": NaN,
  "side": "right",
  "state": "down",
}
`;

exports[`charater class should test the KeyboardEvent keydown 2`] = `
Object {
  "items": Array [],
  "position": NaN,
  "side": "right",
  "state": "run",
}
`;

exports[`charater class should test the KeyboardEvent keydown 3`] = `
Object {
  "items": Array [],
  "position": NaN,
  "side": "left",
  "state": "run",
}
`;

exports[`charater class should test the KeyboardEvent keydown 4`] = `
Object {
  "items": Array [],
  "position": NaN,
  "side": "left",
  "state": "jump",
}
`;

exports[`charater class should test the KeyboardEvent keydown 5`] = `
Object {
  "items": Array [
    Object {
      "element": <div
        class="kunai left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
  ],
  "position": NaN,
  "side": "left",
  "state": "throw",
}
`;

exports[`charater class should test the KeyboardEvent keydown 6`] = `
Object {
  "items": Array [
    Object {
      "element": <div
        class="kunai left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
    Object {
      "element": <div
        class="shuriken left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
  ],
  "position": NaN,
  "side": "left",
  "state": "throw",
}
`;

exports[`charater class should test the KeyboardEvent keyup 1`] = `
Object {
  "items": Array [
    Object {
      "element": <div
        class="kunai left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
    Object {
      "element": <div
        class="shuriken left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
  ],
  "position": NaN,
  "side": "left",
  "state": "stand",
}
`;

exports[`charater class should test the KeyboardEvent keyup 2`] = `
Object {
  "items": Array [
    Object {
      "element": <div
        class="kunai left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
    Object {
      "element": <div
        class="shuriken left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
  ],
  "position": NaN,
  "side": "left",
  "state": "stand",
}
`;

exports[`charater class should test the KeyboardEvent keyup 3`] = `
Object {
  "items": Array [
    Object {
      "element": <div
        class="kunai left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
    Object {
      "element": <div
        class="shuriken left"
        style="position: absolute;"
      />,
      "id": 1644182219303,
      "position": NaN,
    },
  ],
  "position": NaN,
  "side": "left",
  "state": "stand",
}
`;
